common/io.hpp

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_IO_HPP
+#define MENDER_COMMON_IO_HPP
+
+#include <common/error.hpp>
+#include <common/expected.hpp>
+
+#include <system_error>
+#include <vector>
+
+namespace mender::common::io {
+
+using namespace std;
+
+using mender::common::expected::Expected;
+
+class Reader {
+public:
+	virtual ~Reader() {};
+
+	virtual Expected<int64_t, error_condition> Read(vector<byte> &dst) = 0;
+};
+
+class Writer {
+public:
+	virtual ~Writer() {};
+
+	virtual Expected<int64_t, error_condition> Write(const vector<byte> &dst) = 0;
+};
+
+class Closer {
+public:
+	virtual ~Closer() {};
+
+	virtual error_condition Close() = 0;
+};
+
+class ReadWriter : virtual public Reader, virtual public Writer {};
+class ReadCloser : virtual public Reader, virtual public Closer {};
+class WriteCloser : virtual public Reader, virtual public Closer {};
+class ReadWriteCloser : virtual public ReadWriter, virtual public ReadCloser, virtual public WriteCloser {};
```

Comment:
So far I've only used `Closer` from here, but I think it makes sense to have all, for later use.

Reply:
Examples of using all would be nice. Plus why not start adding the surrounding infrastructure also? So something to shuffle bytes from a reader to a writer? Util no-op wrappers?

Reply:
Sure, I can add the trusted and loved `Copy`!

Reply:
TBH, I don't understand any of this. :sweat_smile: